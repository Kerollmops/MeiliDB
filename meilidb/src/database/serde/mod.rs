use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};
use std::error::Error;
use std::fmt;

use serde::ser;

macro_rules! forward_to_unserializable_type {
    ($($ty:ident => $se_method:ident,)*) => {
        $(
            fn $se_method(self, _v: $ty) -> Result<Self::Ok, Self::Error> {
                Err(SerializerError::UnserializableType { name: "$ty" })
            }
        )*
    }
}

pub mod find_id;
pub mod key_to_string;
pub mod value_to_number;
pub mod serializer;
pub mod indexer_serializer;
pub mod deserializer;

pub fn calculate_hash<T: Hash>(t: &T) -> u64 {
    let mut s = DefaultHasher::new();
    t.hash(&mut s);
    s.finish()
}

#[derive(Debug)]
pub enum SerializerError {
    DocumentIdNotFound,
    UnserializableType { name: &'static str },
    Custom(String),
}

impl ser::Error for SerializerError {
    fn custom<T: fmt::Display>(msg: T) -> Self {
        SerializerError::Custom(msg.to_string())
    }
}

impl fmt::Display for SerializerError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            SerializerError::DocumentIdNotFound => {
                write!(f, "serialized document does not have an id according to the schema")
            }
            SerializerError::UnserializableType { name } => {
                write!(f, "Only struct and map types are considered valid documents and
                           can be serialized, not {} types directly.", name)
            },
            SerializerError::Custom(s) => f.write_str(&s),
        }
    }
}

impl Error for SerializerError {}

impl From<String> for SerializerError {
    fn from(value: String) -> SerializerError {
        SerializerError::Custom(value)
    }
}
