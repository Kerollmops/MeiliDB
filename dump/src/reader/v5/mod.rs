//! Here is what a dump v5 look like.
//!
//! ```text
//! .
//! ├── indexes
//! │   ├── 22c269d8-fbbd-4416-bd46-7c7c02849325
//! │   │   ├── documents.jsonl
//! │   │   └── meta.json
//! │   ├── 6d0471ba-2ed1-41de-8ea6-10db10fa2bb8
//! │   │   ├── documents.jsonl
//! │   │   └── meta.json
//! │   └── f7d53ec4-0748-48e6-b66f-1fca9944b0fa
//! │       ├── documents.jsonl
//! │       └── meta.json
//! ├── index_uuids
//! │   └── data.jsonl
//! ├── instance-uid
//! ├── keys
//! ├── metadata.json
//! └── updates
//!     ├── data.jsonl
//!     └── updates_files
//!         └── c83a004a-da98-4b94-b245-3256266c7281
//! ```
//!
//! Here is what `index_uuids/data.jsonl` looks like;
//!
//! ```json
//! {"uid":"dnd_spells","index_meta":{"uuid":"22c269d8-fbbd-4416-bd46-7c7c02849325","creation_task_id":9}}
//! {"uid":"movies","index_meta":{"uuid":"6d0471ba-2ed1-41de-8ea6-10db10fa2bb8","creation_task_id":1}}
//! {"uid":"products","index_meta":{"uuid":"f7d53ec4-0748-48e6-b66f-1fca9944b0fa","creation_task_id":4}}
//! ```
//!

use std::{
    fs::{self, File},
    io::{BufRead, BufReader},
    path::Path,
};

use serde::{Deserialize, Serialize};
use tempfile::TempDir;
use time::OffsetDateTime;
use uuid::Uuid;

use crate::{IndexMetadata, Result, Version};

use super::{compat::v5_to_v6::CompatV5ToV6, DumpReader, IndexReader};

mod keys;
mod meta;
mod settings;
mod tasks;

pub use keys::*;
pub use meta::*;
pub use settings::*;
pub use tasks::*;

pub type Document = serde_json::Map<String, serde_json::Value>;
pub type UpdateFile = File;

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Metadata {
    db_version: String,
    index_db_size: usize,
    update_db_size: usize,
    #[serde(with = "time::serde::rfc3339")]
    dump_date: OffsetDateTime,
}

pub struct V5Reader {
    dump: TempDir,
    metadata: Metadata,
    tasks: BufReader<File>,
    keys: BufReader<File>,
    index_uuid: Vec<IndexUuid>,
}

impl V5Reader {
    pub fn open(dump: TempDir) -> Result<Self> {
        let meta_file = fs::read(dump.path().join("metadata.json"))?;
        let metadata = serde_json::from_reader(&*meta_file)?;
        let index_uuid = File::open(dump.path().join("index_uuids/data.jsonl"))?;
        let index_uuid = BufReader::new(index_uuid);
        let index_uuid = index_uuid
            .lines()
            .map(|line| -> Result<_> { Ok(serde_json::from_str(&line?)?) })
            .collect::<Result<Vec<_>>>()?;

        Ok(V5Reader {
            metadata,
            tasks: BufReader::new(
                File::open(dump.path().join("updates").join("data.jsonl")).unwrap(),
            ),
            keys: BufReader::new(File::open(dump.path().join("keys"))?),
            index_uuid,
            dump,
        })
    }

    pub fn to_v6(self) -> CompatV5ToV6 {
        CompatV5ToV6::new(self)
    }

    pub fn version(&self) -> Version {
        Version::V5
    }

    pub fn date(&self) -> Option<OffsetDateTime> {
        Some(self.metadata.dump_date)
    }

    pub fn instance_uid(&self) -> Result<Option<Uuid>> {
        let uuid = fs::read_to_string(self.dump.path().join("instance-uid"))?;
        Ok(Some(Uuid::parse_str(&uuid)?))
    }

    pub fn indexes(&self) -> Result<impl Iterator<Item = Result<V5IndexReader>> + '_> {
        Ok(self.index_uuid.iter().map(|index| -> Result<_> {
            Ok(V5IndexReader::new(
                index.uid.clone(),
                &self
                    .dump
                    .path()
                    .join("indexes")
                    .join(index.index_meta.uuid.to_string()),
            )?)
        }))
    }

    pub fn tasks(&mut self) -> impl Iterator<Item = Result<(Task, Option<UpdateFile>)>> + '_ {
        (&mut self.tasks).lines().map(|line| -> Result<_> {
            let task: Task = serde_json::from_str(&line?)?;
            if !task.is_finished() {
                if let Some(uuid) = task.get_content_uuid() {
                    let update_file_path = self
                        .dump
                        .path()
                        .join("updates")
                        .join("updates_files")
                        .join(uuid.to_string());
                    Ok((task, Some(File::open(update_file_path).unwrap())))
                } else {
                    Ok((task, None))
                }
            } else {
                Ok((task, None))
            }
        })
    }

    pub fn keys(&mut self) -> impl Iterator<Item = Result<Key>> + '_ {
        (&mut self.keys)
            .lines()
            .map(|line| -> Result<_> { Ok(serde_json::from_str(&line?)?) })
    }
}

pub struct V5IndexReader {
    metadata: IndexMetadata,
    settings: Settings<Checked>,

    documents: BufReader<File>,
}

impl V5IndexReader {
    pub fn new(name: String, path: &Path) -> Result<Self> {
        let meta = File::open(path.join("meta.json"))?;
        let meta: DumpMeta = serde_json::from_reader(meta)?;

        let metadata = IndexMetadata {
            uid: name,
            primary_key: meta.primary_key,
            // FIXME: Iterate over the whole task queue to find the creation and last update date.
            created_at: OffsetDateTime::now_utc(),
            updated_at: OffsetDateTime::now_utc(),
        };

        let ret = V5IndexReader {
            metadata,
            settings: meta.settings.check(),
            documents: BufReader::new(File::open(path.join("documents.jsonl"))?),
        };

        Ok(ret)
    }

    pub fn metadata(&self) -> &IndexMetadata {
        &self.metadata
    }

    pub fn documents(&mut self) -> Result<impl Iterator<Item = Result<Document>> + '_> {
        Ok((&mut self.documents)
            .lines()
            .map(|line| -> Result<_> { Ok(serde_json::from_str(&line?)?) }))
    }

    pub fn settings(&mut self) -> Result<Settings<Checked>> {
        Ok(self.settings.clone())
    }
}

#[cfg(test)]
pub(crate) mod test {
    use std::{fs::File, io::BufReader};

    use flate2::bufread::GzDecoder;
    use tempfile::TempDir;

    use super::*;

    #[test]
    fn read_dump_v5() {
        let dump = File::open("tests/assets/v5.dump").unwrap();
        let dir = TempDir::new().unwrap();
        let mut dump = BufReader::new(dump);
        let gz = GzDecoder::new(&mut dump);
        let mut archive = tar::Archive::new(gz);
        archive.unpack(dir.path()).unwrap();

        let mut dump = V5Reader::open(dir).unwrap();

        // top level infos
        insta::assert_display_snapshot!(dump.date().unwrap(), @"2022-10-04 15:55:10.344982459 +00:00:00");
        insta::assert_display_snapshot!(dump.instance_uid().unwrap().unwrap(), @"9e15e977-f2ae-4761-943f-1eaf75fd736d");

        // tasks
        let tasks = dump.tasks().collect::<Result<Vec<_>>>().unwrap();
        let (tasks, update_files): (Vec<_>, Vec<_>) = tasks.into_iter().unzip();
        insta::assert_json_snapshot!(tasks);
        assert_eq!(update_files.len(), 22);
        assert!(update_files[0].is_none()); // the dump creation
        assert!(update_files[1].is_some()); // the enqueued document addition
        assert!(update_files[2..].iter().all(|u| u.is_none())); // everything already processed

        // keys
        let keys = dump.keys().collect::<Result<Vec<_>>>().unwrap();
        insta::assert_json_snapshot!(keys);

        // indexes
        let mut indexes = dump.indexes().unwrap().collect::<Result<Vec<_>>>().unwrap();
        // the index are not ordered in any way by default
        indexes.sort_by_key(|index| index.metadata().uid.to_string());

        let mut products = indexes.pop().unwrap();
        let mut movies = indexes.pop().unwrap();
        let mut spells = indexes.pop().unwrap();
        assert!(indexes.is_empty());

        // products
        insta::assert_json_snapshot!(products.metadata(), { ".createdAt" => "[now]", ".updatedAt" => "[now]" }, @r###"
        {
          "uid": "products",
          "primaryKey": "sku",
          "createdAt": "[now]",
          "updatedAt": "[now]"
        }
        "###);

        insta::assert_debug_snapshot!(products.settings());
        let documents = products
            .documents()
            .unwrap()
            .collect::<Result<Vec<_>>>()
            .unwrap();
        assert_eq!(documents.len(), 10);
        insta::assert_json_snapshot!(documents);

        // movies
        insta::assert_json_snapshot!(movies.metadata(), { ".createdAt" => "[now]", ".updatedAt" => "[now]" }, @r###"
        {
          "uid": "movies",
          "primaryKey": "id",
          "createdAt": "[now]",
          "updatedAt": "[now]"
        }
        "###);

        insta::assert_debug_snapshot!(movies.settings());
        let documents = movies
            .documents()
            .unwrap()
            .collect::<Result<Vec<_>>>()
            .unwrap();
        assert_eq!(documents.len(), 200);
        insta::assert_debug_snapshot!(documents);

        // spells
        insta::assert_json_snapshot!(spells.metadata(), { ".createdAt" => "[now]", ".updatedAt" => "[now]" }, @r###"
        {
          "uid": "dnd_spells",
          "primaryKey": "index",
          "createdAt": "[now]",
          "updatedAt": "[now]"
        }
        "###);

        insta::assert_debug_snapshot!(spells.settings());
        let documents = spells
            .documents()
            .unwrap()
            .collect::<Result<Vec<_>>>()
            .unwrap();
        assert_eq!(documents.len(), 10);
        insta::assert_json_snapshot!(documents);
    }
}
