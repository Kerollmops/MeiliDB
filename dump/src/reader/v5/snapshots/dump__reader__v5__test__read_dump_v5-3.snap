---
source: dump/src/reader/v5/mod.rs
expression: tasks
---
[
  {
    "id": 21,
    "content": {
      "Dump": {
        "uid": "20221004-155510279"
      }
    },
    "events": [
      {
        "Created": "2022-10-04T15:55:10.281165892Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-04T15:55:10.340501133Z",
          "batch_id": 21
        }
      },
      {
        "Processing": "2022-10-04T15:55:10.340507253Z"
      }
    ]
  },
  {
    "id": 20,
    "content": {
      "DocumentAddition": {
        "index_uid": "movies_2",
        "content_uuid": "c83a004a-da98-4b94-b245-3256266c7281",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": null,
        "documents_count": 200,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-04T15:55:10.272276202Z"
      }
    ]
  },
  {
    "id": 19,
    "content": {
      "DocumentAddition": {
        "index_uid": "movies",
        "content_uuid": "eec3f3ed-cbc9-4b29-95d1-99e7224b1c18",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": null,
        "documents_count": 100,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-04T15:55:10.259722791Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-04T15:55:10.273271829Z",
          "batch_id": 19
        }
      },
      {
        "Processing": "2022-10-04T15:55:10.273278199Z"
      },
      {
        "Succeeded": {
          "result": {
            "DocumentAddition": {
              "indexed_documents": 100
            }
          },
          "timestamp": "2022-10-04T15:55:10.33561842Z"
        }
      }
    ]
  },
  {
    "id": 18,
    "content": {
      "DocumentAddition": {
        "index_uid": "dnd_spells",
        "content_uuid": "b9a17abc-6034-497c-a09d-2295bfadebb8",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": "index",
        "documents_count": 10,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-04T15:55:02.364638737Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-04T15:55:02.37723066Z",
          "batch_id": 18
        }
      },
      {
        "Processing": "2022-10-04T15:55:02.37723266Z"
      },
      {
        "Succeeded": {
          "result": {
            "DocumentAddition": {
              "indexed_documents": 10
            }
          },
          "timestamp": "2022-10-04T15:55:02.394503431Z"
        }
      }
    ]
  },
  {
    "id": 17,
    "content": {
      "DocumentAddition": {
        "index_uid": "dnd_spells",
        "content_uuid": "3f905db8-3052-4a01-a52d-440d8c5d8333",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": null,
        "documents_count": 10,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-04T15:55:02.116014762Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-04T15:55:02.128681906Z",
          "batch_id": 17
        }
      },
      {
        "Processing": "2022-10-04T15:55:02.128683566Z"
      },
      {
        "Succeeded": {
          "result": {
            "DocumentAddition": {
              "indexed_documents": 10
            }
          },
          "timestamp": "2022-10-04T15:55:02.145816865Z"
        }
      }
    ]
  },
  {
    "id": 16,
    "content": {
      "DocumentAddition": {
        "index_uid": "products",
        "content_uuid": "a7d5925b-460d-4331-b11f-c8639fa6629c",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": "sku",
        "documents_count": 10,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-04T15:55:01.867101853Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-04T15:55:01.879802098Z",
          "batch_id": 16
        }
      },
      {
        "Processing": "2022-10-04T15:55:01.879803378Z"
      },
      {
        "Succeeded": {
          "result": {
            "DocumentAddition": {
              "indexed_documents": 10
            }
          },
          "timestamp": "2022-10-04T15:55:01.897325373Z"
        }
      }
    ]
  },
  {
    "id": 15,
    "content": {
      "DocumentAddition": {
        "index_uid": "products",
        "content_uuid": "15ba7b99-e0a6-4376-be83-1e4f1ada4613",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": null,
        "documents_count": 10,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-04T15:55:01.245884922Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-04T15:55:01.258597317Z",
          "batch_id": 15
        }
      },
      {
        "Processing": "2022-10-04T15:55:01.258598227Z"
      },
      {
        "Succeeded": {
          "result": {
            "DocumentAddition": {
              "indexed_documents": 10
            }
          },
          "timestamp": "2022-10-04T15:55:01.270564108Z"
        }
      }
    ]
  },
  {
    "id": 14,
    "content": {
      "SettingsUpdate": {
        "index_uid": "products",
        "settings": {
          "displayedAttributes": "NotSet",
          "searchableAttributes": "NotSet",
          "filterableAttributes": "NotSet",
          "sortableAttributes": "NotSet",
          "rankingRules": "NotSet",
          "stopWords": "NotSet",
          "synonyms": {
            "Set": {
              "android": [
                "phone",
                "smartphone"
              ],
              "iphone": [
                "phone",
                "smartphone"
              ],
              "phone": [
                "smartphone",
                "iphone",
                "android"
              ]
            }
          },
          "distinctAttribute": "NotSet",
          "typoTolerance": "NotSet",
          "faceting": "NotSet",
          "pagination": "NotSet"
        },
        "is_deletion": false,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-04T15:55:00.630109164Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-04T15:55:00.643607641Z",
          "batch_id": 14
        }
      },
      {
        "Processing": "2022-10-04T15:55:00.643609011Z"
      },
      {
        "Succeeded": {
          "result": "Other",
          "timestamp": "2022-10-04T15:55:00.654385523Z"
        }
      }
    ]
  },
  {
    "id": 13,
    "content": {
      "SettingsUpdate": {
        "index_uid": "movies",
        "settings": {
          "displayedAttributes": "NotSet",
          "searchableAttributes": "NotSet",
          "filterableAttributes": "NotSet",
          "sortableAttributes": "NotSet",
          "rankingRules": {
            "Set": [
              "words",
              "typo",
              "proximity",
              "attribute",
              "sort",
              "exactness",
              "release_date:asc"
            ]
          },
          "stopWords": "NotSet",
          "synonyms": "NotSet",
          "distinctAttribute": "NotSet",
          "typoTolerance": "NotSet",
          "faceting": "NotSet",
          "pagination": "NotSet"
        },
        "is_deletion": false,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-04T15:55:00.412114406Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-04T15:55:00.425715225Z",
          "batch_id": 13
        }
      },
      {
        "Processing": "2022-10-04T15:55:00.425716285Z"
      },
      {
        "Succeeded": {
          "result": "Other",
          "timestamp": "2022-10-04T15:55:00.436408825Z"
        }
      }
    ]
  },
  {
    "id": 12,
    "content": {
      "SettingsUpdate": {
        "index_uid": "movies",
        "settings": {
          "displayedAttributes": "NotSet",
          "searchableAttributes": "NotSet",
          "filterableAttributes": {
            "Set": [
              "genres",
              "id"
            ]
          },
          "sortableAttributes": {
            "Set": [
              "release_date"
            ]
          },
          "rankingRules": "NotSet",
          "stopWords": "NotSet",
          "synonyms": "NotSet",
          "distinctAttribute": "NotSet",
          "typoTolerance": "NotSet",
          "faceting": "NotSet",
          "pagination": "NotSet"
        },
        "is_deletion": false,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-04T15:55:00.18896188Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-04T15:55:00.207803798Z",
          "batch_id": 12
        }
      },
      {
        "Processing": "2022-10-04T15:55:00.207804798Z"
      },
      {
        "Succeeded": {
          "result": "Other",
          "timestamp": "2022-10-04T15:55:00.218448708Z"
        }
      }
    ]
  },
  {
    "id": 11,
    "content": {
      "DocumentAddition": {
        "index_uid": "movies",
        "content_uuid": "f6fc7f47-f46d-4161-aabb-5b9909c175e7",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": null,
        "documents_count": 10,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-04T15:54:59.971297669Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-04T15:54:59.984797797Z",
          "batch_id": 11
        }
      },
      {
        "Processing": "2022-10-04T15:54:59.984799097Z"
      },
      {
        "Succeeded": {
          "result": {
            "DocumentAddition": {
              "indexed_documents": 10
            }
          },
          "timestamp": "2022-10-04T15:55:00.002941623Z"
        }
      }
    ]
  },
  {
    "id": 10,
    "content": {
      "DocumentAddition": {
        "index_uid": "movies",
        "content_uuid": "f86fa59e-8557-4057-8fc0-212b2218746f",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": null,
        "documents_count": 100,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-04T15:51:44.147743385Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-04T15:51:44.458468546Z",
          "batch_id": 10
        }
      },
      {
        "Processing": "2022-10-04T15:51:44.458473756Z"
      },
      {
        "Succeeded": {
          "result": {
            "DocumentAddition": {
              "indexed_documents": 100
            }
          },
          "timestamp": "2022-10-04T15:51:44.825304055Z"
        }
      }
    ]
  },
  {
    "id": 9,
    "content": {
      "DocumentAddition": {
        "index_uid": "movies",
        "content_uuid": "d0a0e14c-c668-4fc7-866d-5bac5bdb563a",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": null,
        "documents_count": 90,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-04T15:51:44.142779621Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-04T15:51:44.14804443Z",
          "batch_id": 9
        }
      },
      {
        "Processing": "2022-10-04T15:51:44.14805041Z"
      },
      {
        "Succeeded": {
          "result": {
            "DocumentAddition": {
              "indexed_documents": 90
            }
          },
          "timestamp": "2022-10-04T15:51:44.453183668Z"
        }
      }
    ]
  },
  {
    "id": 8,
    "content": {
      "Dump": {
        "uid": "20221004-155144042"
      }
    },
    "events": [
      {
        "Created": "2022-10-04T15:51:44.042750953Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-04T15:51:44.056655459Z",
          "batch_id": 8
        }
      },
      {
        "Processing": "2022-10-04T15:51:44.056661399Z"
      },
      {
        "Succeeded": {
          "result": "Other",
          "timestamp": "2022-10-04T15:51:44.078996365Z"
        }
      }
    ]
  },
  {
    "id": 7,
    "content": {
      "DocumentAddition": {
        "index_uid": "dnd_spells",
        "content_uuid": "50db5436-0812-4047-a3b7-cf0278e7a9fd",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": "index",
        "documents_count": 10,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-04T15:51:37.628596827Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-04T15:51:37.638649517Z",
          "batch_id": 7
        }
      },
      {
        "Processing": "2022-10-04T15:51:37.638650617Z"
      },
      {
        "Succeeded": {
          "result": {
            "DocumentAddition": {
              "indexed_documents": 10
            }
          },
          "timestamp": "2022-10-04T15:51:37.67796098Z"
        }
      }
    ]
  },
  {
    "id": 6,
    "content": {
      "DocumentAddition": {
        "index_uid": "dnd_spells",
        "content_uuid": "06201cce-3d0c-4c4f-85c3-a4eccb505b62",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": null,
        "documents_count": 10,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-04T15:51:37.381094632Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-04T15:51:37.394320725Z",
          "batch_id": 6
        }
      },
      {
        "Processing": "2022-10-04T15:51:37.394321835Z"
      },
      {
        "Failed": {
          "error": {
            "message": "The primary key inference process failed because the engine did not find any fields containing `id` substring in their name. If your document identifier does not contain any `id` substring, you can set the primary key of the index.",
            "code": "primary_key_inference_failed",
            "type": "invalid_request",
            "link": "https://docs.meilisearch.com/errors#primary_key_inference_failed"
          },
          "timestamp": "2022-10-04T15:51:37.522389886Z"
        }
      }
    ]
  },
  {
    "id": 5,
    "content": {
      "DocumentAddition": {
        "index_uid": "products",
        "content_uuid": "3a38827b-6c9a-41ac-9a19-979bc19a32aa",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": "sku",
        "documents_count": 10,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-04T15:51:37.132755272Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-04T15:51:37.146030327Z",
          "batch_id": 5
        }
      },
      {
        "Processing": "2022-10-04T15:51:37.146031377Z"
      },
      {
        "Succeeded": {
          "result": {
            "DocumentAddition": {
              "indexed_documents": 10
            }
          },
          "timestamp": "2022-10-04T15:51:37.167878265Z"
        }
      }
    ]
  },
  {
    "id": 4,
    "content": {
      "DocumentAddition": {
        "index_uid": "products",
        "content_uuid": "9696c1a6-4ff9-4ce9-8c68-06b6daa5e0dc",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": null,
        "documents_count": 10,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-04T15:51:36.53155275Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-04T15:51:36.544818974Z",
          "batch_id": 4
        }
      },
      {
        "Processing": "2022-10-04T15:51:36.544820074Z"
      },
      {
        "Failed": {
          "error": {
            "message": "The primary key inference process failed because the engine did not find any fields containing `id` substring in their name. If your document identifier does not contain any `id` substring, you can set the primary key of the index.",
            "code": "primary_key_inference_failed",
            "type": "invalid_request",
            "link": "https://docs.meilisearch.com/errors#primary_key_inference_failed"
          },
          "timestamp": "2022-10-04T15:51:36.550259665Z"
        }
      }
    ]
  },
  {
    "id": 3,
    "content": {
      "SettingsUpdate": {
        "index_uid": "products",
        "settings": {
          "displayedAttributes": "NotSet",
          "searchableAttributes": "NotSet",
          "filterableAttributes": "NotSet",
          "sortableAttributes": "NotSet",
          "rankingRules": "NotSet",
          "stopWords": "NotSet",
          "synonyms": {
            "Set": {
              "android": [
                "phone",
                "smartphone"
              ],
              "iphone": [
                "phone",
                "smartphone"
              ],
              "phone": [
                "smartphone",
                "iphone",
                "android"
              ]
            }
          },
          "distinctAttribute": "NotSet",
          "typoTolerance": "NotSet",
          "faceting": "NotSet",
          "pagination": "NotSet"
        },
        "is_deletion": false,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-04T15:51:35.939396731Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-04T15:51:35.952669434Z",
          "batch_id": 3
        }
      },
      {
        "Processing": "2022-10-04T15:51:35.952670314Z"
      },
      {
        "Succeeded": {
          "result": "Other",
          "timestamp": "2022-10-04T15:51:36.087768554Z"
        }
      }
    ]
  },
  {
    "id": 2,
    "content": {
      "SettingsUpdate": {
        "index_uid": "movies",
        "settings": {
          "displayedAttributes": "NotSet",
          "searchableAttributes": "NotSet",
          "filterableAttributes": "NotSet",
          "sortableAttributes": "NotSet",
          "rankingRules": {
            "Set": [
              "words",
              "typo",
              "proximity",
              "attribute",
              "sort",
              "exactness",
              "release_date:asc"
            ]
          },
          "stopWords": "NotSet",
          "synonyms": "NotSet",
          "distinctAttribute": "NotSet",
          "typoTolerance": "NotSet",
          "faceting": "NotSet",
          "pagination": "NotSet"
        },
        "is_deletion": false,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-04T15:51:35.721766758Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-04T15:51:35.735028862Z",
          "batch_id": 2
        }
      },
      {
        "Processing": "2022-10-04T15:51:35.735030412Z"
      },
      {
        "Succeeded": {
          "result": "Other",
          "timestamp": "2022-10-04T15:51:35.745160333Z"
        }
      }
    ]
  },
  {
    "id": 1,
    "content": {
      "SettingsUpdate": {
        "index_uid": "movies",
        "settings": {
          "displayedAttributes": "NotSet",
          "searchableAttributes": "NotSet",
          "filterableAttributes": {
            "Set": [
              "genres",
              "id"
            ]
          },
          "sortableAttributes": {
            "Set": [
              "release_date"
            ]
          },
          "rankingRules": "NotSet",
          "stopWords": "NotSet",
          "synonyms": "NotSet",
          "distinctAttribute": "NotSet",
          "typoTolerance": "NotSet",
          "faceting": "NotSet",
          "pagination": "NotSet"
        },
        "is_deletion": false,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-04T15:51:35.50733827Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-04T15:51:35.517545782Z",
          "batch_id": 1
        }
      },
      {
        "Processing": "2022-10-04T15:51:35.517547002Z"
      },
      {
        "Succeeded": {
          "result": "Other",
          "timestamp": "2022-10-04T15:51:35.539766447Z"
        }
      }
    ]
  },
  {
    "id": 0,
    "content": {
      "DocumentAddition": {
        "index_uid": "movies",
        "content_uuid": "8b328d9d-ba52-4e16-b22d-e45145733ae0",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": null,
        "documents_count": 10,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-04T15:51:35.291992167Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-04T15:51:35.302592417Z",
          "batch_id": 0
        }
      },
      {
        "Processing": "2022-10-04T15:51:35.302593877Z"
      },
      {
        "Succeeded": {
          "result": {
            "DocumentAddition": {
              "indexed_documents": 10
            }
          },
          "timestamp": "2022-10-04T15:51:35.459210749Z"
        }
      }
    ]
  }
]
